### DISCUSSION TITLE
Processing power and Sample Rate, Sample Size

### DISCUSSION THREAD

**Discussion by MichaelPNolan** [marcel-licence/esp32_basic_synth#53] [Show and tell]
In config.h
some comments - I have this PCM5102 DAC that is 32bit capable and 384khz ... but lets get real. ESP32 ... its not god CPU power SOC. Especially if you are toying around and its perhaps doing a few other jobs - like using the USB and other midi.

Symptom - bad sounds, ghost notes, doesn't sound like the waveform - unpleasant sound.
Solution: 1) Make sure you have your ESP32 on 240mhz (assume the IDE is setting that using the menu to set clock)
               2) 16bit (like you see Marcel in videos using i assume) and 22khz sample rate. Finally it sounds how I believe it should sound - and that is very sweet and rewarding. I'm so happy. Now I'm off to build my own USB midi controller since I have to wait for another thing in the mail. I'll use one of these Mega32U4 pro micro things - apparently they easily implement midi device (as opposed to the Midi Host for this synth module.

Any other crucial make it or break it tips?
I'll make a video and show where I'm at.
:-)
```
/*
 * You can modify the sample rate as you want
 */

#ifdef ESP32_AUDIO_KIT
#define SAMPLE_RATE 44100
#define SAMPLE_SIZE_16BIT
#else
#define SAMPLE_RATE 22000  //even 32khz - kind of sucks and you get gliches and wrong sustained notes and fails
#define SAMPLE_SIZE_16BIT //previous default 32bit is possible but go with 16bit
//#define SAMPLE_SIZE_8BIT  //default experimental
#endif
```

**Comment by marcel-licence** [comment]
Thanks for sharing. One step which could be always useful is just to generate a sine to verify the DAC creates the correct signal. I am running my code with 48KHz on the ESP32 when using the PCM5102 DAC. I am not sure how sensitive the USB host breakout is to electrical noise which could be also a problem on breadboards. I had issues with MIDI via the opto coupler. After soldering it to a separeted board it works well. The code is intentionally written that you will get audible noises in case you are doing too much in the loop process. You also should avoid calling any function within this loop which may take longer (for example doing SPI).

**Comment by MichaelPNolan** [comment]
Yes - believe that I can possibly tune the upkeep in the loop to be less frequent and see how much more sample frequency is working. I'm just relieved that all the things went away when I slowed it down - there were all these ghost notes. So it made me wonder about the code and settings but it vanished.
As for noise - when I solder it together I'll put a 5v power isolator in which I've used in my amplifier/bluetooth and pre-amp situations where the bluetooth was audio (its dac etc) had noise. The ESP32 is certainly generating a lot of noise on the earth. There are some tricks like a 10ohm resistor to ground that some people use when joining audio systems together to break loops. I've seen that but its not so common so i'm careful with that. 10 ohms isn't much resistance though.

I've seen someone have issues with the optocoupler set the resistor pull up to +v to 560 ohm and it seemed to reduce delay comms issues vs the 10k resistor they first used. I've not set up any midi using serial connection to microcontroller connections or optocoupler yet.

**Reply by marcel-licence** [reply]
To prevent ground/USB noise I am using a "3.5MM Jack Audio Loop Noise Isolator Signal Anti-interference Filter". I didn't check how much the frequency response will be influenced but for recording it's much better than without.

For MIDI I tried now to use the H11L1 and the schematic from here: https://www.elektormagazine.de/articles/breakout-board-fur-midi-i-o-mit-din-oder-klinkenbuchsen but without coils etc. Just 220 ohms resistor in series with the input.
H11L1 is directly connected to the supply (3.3V) with parallel 100nF capacity. The output is pulled up by 1k (maybe check the settings in the code the internal ESP32 pullup may also be active).

**Reply by MichaelPNolan** [reply]
Sometimes I use transformer isolation for my audio - especially new things connected to amps and pre-amps I'm testing. Thanks for the link.
